### DISCUSSION TITLE
Compuound shapes

### DISCUSSION THREAD

**Discussion by albinaask** [godot-jolt/godot-jolt#656] [Questions]
Hi. is there any plans for integrating the Jolt compound shapes to this port?

Thanks in advance?

**Comment by mihe** [comment]
Jolt's `StaticCompoundShape` is used already whenever you have more than one shape attached to a body.

There are no plans to expose the different compound shapes as their own nodes or anything like that.

**Reply by albinaask** [reply]
Hmmm, I see! The problem that I'm having is that when I add many complex shapes to a body (in the realm of 15k or so), the simulations have weird hiccups on the main thread. Even though I only have a single body in the world, so I wouldn't think it's collision related at all. It's like it recalculates the internal BVH or something. The same happends when I have a sparse grid of many bodies with many shapes, again without any collisions. It's like the performance is bound by the number of shapes... I thought it was since the StaticCompoundShape wasn't implemented...But now I'm not sure. Other Ideas I've had have been that it is a scene tree issue since I'm loading the scene tree with 30k nodes(one Shape & one Meshinstance per shape in the body). But if that was the problem I would suspect the slowness would occur every frame... As opposed to now when it only does so once every like three seconds or so... Have you any idea of what could cause this behavior?

**Reply by mihe** [reply]
> The problem that I'm having is that when I add many complex shapes to a body (in the realm of 15k or so), the simulations have weird hiccups on the main thread.

While you're adding the shapes or afterwards?

> It's like it recalculates the internal BVH or something.

It does indeed do this whenever you add a shape, as I currently rebuild the entire `StaticCompoundShape` from scratch on any modification of any shape, including changes to shape properties, but once it's built that shouldn't happen.

> Have you any idea of what could cause this behavior?

It's hard to say without being able to profile it. If you're able to put together a representative minimal repro project then feel free to create an issue and upload it there and I can take a look.
